### ISSUE TITLE
SfM with combination of known and unknown camera poses

### ISSUE THREAD

**Issue by manurare** [openMVG/openMVG#1892]
Hi,

I have generated some spherical images with known camera poses in Blender. Apart from these, I have additional panoramas from the same scene but with unknown camera poses. I would like to find the camera parameters for these images leveraging the known parameters. Eventually, the returned parameters shall be aligned in the same scale as the ones from Blender. Is this possible?

I have read the documentation on `openMVG_main_ComputeStructureFromKnownPoses ` but it looks like that is not exactly what I am trying to do as this doesn't return poses but geometry since all camera poses are known. Really what I am looking for is a way of adding to the known camera poses more images from which to infer their parameters. 

Thanks!

**Comment by manurare** [comment]
Reading a bit through the library. I think what I have to do is the following:

1. Create my own sfm data container with my known camera poses
2. Run `main_ComputeFeatures` and `main_ComputeMatches`
3. Run `main_IncrementalSfM2`.

I have several questions though. Does `main_IncrementalSfM2` fix the known camera poses in BA and only optimizes for the unknown? I can't see spherical camera model in `main_IncrementalSfM2`. Is it supported?

**Comment by pmoulon** [comment]
Hello @manurare , Sorry for the late answer.

The best path to do so as you found would be to attempt what you have noted in your second message.

1. You can use SfMInitImageListing and use the spherical model then use python to add the pose information
2. Perform 2 as you noted
3. Perform 3 as noted

main_IncrementalSfM2 is not fixing yet the BA of the input pose, but I plan to add some support for that soon.
The spherical model in `main_IncrementalSfM2` is supported (it will use the camera model defined in the JSON - all your image will be associated to a single Spherical Intrinsic if they have the same image resolution).

You can see pure C++ importer here https://github.com/openMVG/openMVG/tree/develop/src/software/SfM/import

**Comment by manurare** [comment]
Hey.

I am facing a problem though. I have created successfully my own sfm data container which consists on 18 GT spherical poses. I have 21 images therefore after `main_IncrementalSfM2` I should get 21 poses but I can't manage to get those. I am executing the script as follow 

        pRecons = subprocess.Popen( [os.path.join(OPENMVG_SFM_BIN, "openMVG_main_IncrementalSfM2"),  "-i",
                                     input_dir+"../GTPoses_sfm_data.json", "-m", matches_dir, "-o", reconstruction_dir,
                                     "-S", "EXISTING_POSE"] ) 
I am trying to initialize the reconstruction from the existing poses but the result I am getting does not contains 21 poses but 18. If I use `STELLAR` as default it doesn't even perform BA. Am I missing something here?

**Comment by DHNicoles** [comment]
@manurare Hi, I was looking for how to import the known poses of 360camera into the sfm_data.json file, but I couldn't find it. Is there an example of using ComputeStructureFromKnownPoses for a spherical camera?
I have read the source code here, maybe I missed something ？
`io_loadImages.hpp io_readGTETH3D.hpp io_readGTKitti.hpp io_readGTStrecha.hpp io_readGTDTUMVS.hpp io_readGTInterface.hpp io_readGTMiddleBury.hpp main_SfMInit_ImageListingFromKnownPoses.cpp`
Thank you.

**Comment by pmoulon** [comment]
@manurare It is already somehow a good news that you are still be able to keep your initial 18 poses (it shows that you have provided compatible Rotation and Translation information for each camera).
-> Have you computed the features and matches with the sfm_data.json that contains the 21 poses?
-> If yes, it could means that the image/feature matching connectivity is not dense enough to allow a resection of your remaining camera -> Feel free to share your data (images and starting JSON) to us so we can have a look (openmvg-team[AT]googlegroups.com)
-> If you have Graphviz on your machine, openMVG should plot the image connectivity graph and it should show you the linked image after image machine and robust image filtering
-> You can also use openmvg_main_exportTracks/Matches to visualize if your images are connected and see with which keypoint.
-> If you have no keypoint connection -> try to use the feature preset `-p HIGH` -> see if the baseline is big...

Hoping that those info helps. If you are willing to share your dataset, we would be happy to create a tutorial from it for the wiki.


@manurare & @DHNicoles here some info regarding what the input SfM_data.json could look like with your images poses
I ran the command line and the dataset shared here https://github.com/openMVG/Image_datasets/blob/master/360_dataset/README.md and then convert the output sfm_data.bin to a sfm_data.json with just the View, Intrinsic and Extrinsic data:
```
openMVG_main_ConvertSfM_DataFormat.exe -V -I -E -i ./360_dataset\360_school\match\reconstruction\sfm_data.bin -o .\360_dataset\360_school\match\reconstruction\sfm_data_example.json
```

```
{
    "sfm_data_version": "0.3",
    "root_path": "360_dataset\\360_school\\images",
    "views": [
        {
            "key": 0,
            "value": {
                "polymorphic_id": 1073741824,
                "ptr_wrapper": {
                    "id": 2147483649,
                    "data": {
                        "local_path": "",
                        "filename": "R0010939.jpg",
                        "width": 5376,
                        "height": 2688,
                        "id_view": 0,
                        "id_intrinsic": 0,
                        "id_pose": 0
                    }
                }
            }
        },
        {
            "key": 1,
            "value": {
                "polymorphic_id": 1073741824,
                "ptr_wrapper": {
                    "id": 2147483650,
                    "data": {
                        "local_path": "",
                        "filename": "R0010940.jpg",
                        "width": 5376,
                        "height": 2688,
                        "id_view": 1,
                        "id_intrinsic": 0,
                        "id_pose": 1
                    }
                }
            }
        },
        {
            "key": 2,
            "value": {
                "polymorphic_id": 1073741824,
                "ptr_wrapper": {
                    "id": 2147483651,
                    "data": {
                        "local_path": "",
                        "filename": "R0010941.jpg",
                        "width": 5376,
                        "height": 2688,
                        "id_view": 2,
                        "id_intrinsic": 0,
                        "id_pose": 2
                    }
                }
            }
        },
        {
            "key": 3,
            "value": {
                "polymorphic_id": 1073741824,
                "ptr_wrapper": {
                    "id": 2147483652,
                    "data": {
                        "local_path": "",
                        "filename": "R0010942.jpg",
                        "width": 5376,
                        "height": 2688,
                        "id_view": 3,
                        "id_intrinsic": 0,
                        "id_pose": 3
                    }
                }
            }
        }
    ],
    "intrinsics": [
        {
            "key": 0,
            "value": {
                "polymorphic_id": 2147483649,
                "polymorphic_name": "spherical",
                "ptr_wrapper": {
                    "id": 2147483653,
                    "data": {
                        "value0": {
                            "width": 5376,
                            "height": 2688
                        }
                    }
                }
            }
        }
    ],
    "extrinsics": [
        {
            "key": 0,
            "value": {
                "rotation": [
                    [
                        0.9999883270852489,
                        0.004812442654244155,
                        0.0004313802787725093
                    ],
                    [
                        -0.004816592569960024,
                        0.9999363838110018,
                        0.010199449347482596
                    ],
                    [
                        -0.0003822685709135046,
                        -0.01020140807322542,
                        0.9999478912143688
                    ]
                ],
                "center": [
                    -0.026954436886442606,
                    0.010249102993893179,
                    -0.0008088500958622927
                ]
            }
        },
        {
            "key": 1,
            "value": {
                "rotation": [
                    [
                        0.9961163584073676,
                        0.004543401418951389,
                        -0.08792927849579528
                    ],
                    [
                        -0.003644498542215396,
                        0.9999394758225165,
                        0.010380863266933409
                    ],
                    [
                        0.08797112107743453,
                        -0.010020089587286533,
                        0.9960726277039451
                    ]
                ],
                "center": [
                    -0.3388952469957019,
                    0.00947828846791296,
                    -0.05874602685926579
                ]
            }
        },
        {
            "key": 2,
            "value": {
                "rotation": [
                    [
                        0.9907516392844542,
                        -0.0015993112985443633,
                        0.13567841191043285
                    ],
                    [
                        0.0005906713843929864,
                        0.9999718949718197,
                        0.007473979782125521
                    ],
                    [
                        -0.13568655188515328,
                        -0.007324716365724349,
                        0.9907247388491312
                    ]
                ],
                "center": [
                    -0.6370781341976203,
                    0.00831184769860101,
                    -0.09664587827306165
                ]
            }
        },
        {
            "key": 3,
            "value": {
                "rotation": [
                    [
                        0.9674382922904713,
                        -0.012945765755660578,
                        0.2527757064258673
                    ],
                    [
                        0.012569678359278514,
                        0.9999161846411414,
                        0.0031027211681324336
                    ],
                    [
                        -0.2527946870407702,
                        0.00017561805846102378,
                        0.9675199302144926
                    ]
                ],
                "center": [
                    -0.9424480020876315,
                    0.007419223477062246,
                    -0.1608076549267645
                ]
            }
        }
    ],
    "structure": [],
    "control_points": []
}
```

**Comment by DHNicoles** [comment]
@pmoulon 
I just tried to put the SfM_data.json you provided above in the mathes folder, and then execute the pipeline one by one:
```
dataset=../Image_datasets/360_dataset/360_school
dataset_out=./output/360_school
openMVG_main_ComputeFeatures -i $dataset_out/matches/SfM_data.json -o $dataset_out/matches
openMVG_main_ComputeMatches -i $dataset_out/matches/SfM_data.json -o $dataset_out/matches -g a
openMVG_main_ComputeStructureFromKnownPoses -i $dataset_out/matches/sfm_data.json -m $dataset_out/matches -o $dataset_out/reconstruction
```
But the last line of the command **crashed** when it was executed. I read the _openMVG/sfm/sfm_data_filters_frustum.cpp_ source code and found that in the **void Frustum_Filter::initFrustum** function,
```
if (!isPinhole(iterIntrinsic->second->getType()))
  continue; 
```
So, am I doing something wrong? Or what is the correct pipeline to use the known Spherical camera pose?

**Comment by DHNicoles** [comment]
> @pmoulon
> I just tried to put the SfM_data.json you provided above in the mathes folder, and then execute the pipeline one by one:
> 
> ```
> dataset=../Image_datasets/360_dataset/360_school
> dataset_out=./output/360_school
> openMVG_main_ComputeFeatures -i $dataset_out/matches/SfM_data.json -o $dataset_out/matches
> openMVG_main_ComputeMatches -i $dataset_out/matches/SfM_data.json -o $dataset_out/matches -g a
> openMVG_main_ComputeStructureFromKnownPoses -i $dataset_out/matches/sfm_data.json -m $dataset_out/matches -o $dataset_out/reconstruction
> ```
> 
> But the last line of the command **crashed** when it was executed. I read the _openMVG/sfm/sfm_data_filters_frustum.cpp_ source code and found that in the **void Frustum_Filter::initFrustum** function,
> 
> ```
> if (!isPinhole(iterIntrinsic->second->getType()))
>   continue; 
> ```
> 
> Does this mean that the Spherical type in SfM_data.json cannot be used directly?
> So, am I doing something wrong? Or what is the correct pipeline to use the known Spherical camera pose?

I tried to skip openMVG_main_ComputeStructureFromKnownPoses and use openMVG_main_IncrementalSfM2 directly, and it worked!

**Comment by manurare** [comment]
> @manurare It is already somehow a good news that you are still be able to keep your initial 18 poses (it shows that you have provided compatible Rotation and Translation information for each camera).
-> Have you computed the features and matches with the sfm_data.json that contains the 21 poses?
-> If yes, it could means that the image/feature matching connectivity is not dense enough to allow a resection of your remaining camera -> Feel free to share your data (images and starting JSON) to us so we can have a look (openmvg-team[AT]googlegroups.com)
-> If you have Graphviz on your machine, openMVG should plot the image connectivity graph and it should show you the linked image after image machine and robust image filtering
-> You can also use openmvg_main_exportTracks/Matches to visualize if your images are connected and see with which keypoint.
-> If you have no keypoint connection -> try to use the feature preset -p HIGH -> see if the baseline is big...

@pmoulon I managed to get some unknown poses now. The problem was that when creating my sfm data container I wasn't populating the views with the unknown 3 poses. I was hoping that after the `computer_features` and `compute_matches` it would create a 21 poses sfm data which eventually would combine with my 18 initial poses sfm data inside `main_IncrementalSfM2` filling the missing info. Summarizing, my input sfm data has 21 views and 18 poses. After `main_IncrementalSfM2` I get 1 of the 3 missing poses left but I am missing some of the GT poses :(. My unknown poses are the views with ids 6,13 and 20. I got the pose 20 but I miss all the info from 7,8,9,10,11,12 which is whole room in the building of matterport. 

![resectioning](https://user-images.githubusercontent.com/43776720/119116986-b3037480-ba20-11eb-818b-5afd55d4e65d.png)



The feature preset is set in -p HIGH. This is my connectivity graph

![Screenshot from 2021-05-21 10-41-07](https://user-images.githubusercontent.com/43776720/119117430-1ee5dd00-ba21-11eb-922b-169bcec729c7.png)



I am using matterport3D data. Not quite sure if I can share. If it is allowed, I will be happy to share!.

**Comment by pmoulon** [comment]
Thank you for the report.
I would take a look to the feature matches you have.
Perhaps we are not able to register the last images because we don't have 3D-2D correspondences to find the absolute pose.

**Comment by pmoulon** [comment]
Regarding `miss all the info from 7,8,9,10,11,12 which is whole room in the building of matterport.` It could be we are processing only the largest connected component. We can check about that next week

**Comment by pmoulon** [comment]
Hello @manurare 
Made some test on my side, I wonder if there is something wrong on the parameter
Playing with SIFT ULTRA I noticed that the initial triangulation is very bad, but then the structure is coming back...
![image](https://user-images.githubusercontent.com/3235050/122657543-39db6800-d119-11eb-9d18-6d3c1846c777.png)

Else with default settings I got similar results as you.

PS: Since your images are upright you can use the option `-u 1` to speed up the matching

**Comment by manurare** [comment]
Hi @pmoulon. Thanks for taking your time. I think at the time I sent you the poses they were wrong. As I rendered the images from Blender, the poses depict the transformation from blender camera cs to blender world cs. I then realised that the poses should be from openMVG camera cs to blender world. What I did was multiplying the rotation with a transformation from blender cam cs to openMVG camera cs. Still, I dont get good results but at least I solved that bug.
